### DISCUSSION TITLE
MediaPlayer audio inquiry

### DISCUSSION THREAD

**Discussion by leeyunjai82** [aiortc/aiortc#741] [Q&A]
Hi, everyone.
I use i2s mic, raspberrypi camera and raspberrypi.
I want to use my mic and camera in MediaPlayer Function for videocall of janus-gateway  
>> no audio traffic, video ok
Please check if I have set MediaPlayer().audio up wrong

Thank you
'''
    media = {'audio': True, 'video': True}

    vplayer = MediaPlayer("/dev/video0", format='v4l2', options={'framerate':'5', 'video_size':'320x240'})
    if media['video'] and vplayer and vplayer.video:
        pc.addTrack(vplayer.video)
    
    aplayer = MediaPlayer("sysdefault:CARD=sndrpii2scard", format="alsa", options={'channels':'2', 'sample_rate':'44000'})
    if media['audio'] and aplayer and aplayer.audio:
        pc.addTrack(aplayer.audio)
        media['audio'] = True
'''

>>Result
if media['audio'] set True --> no audio traffic
video ok

Please check if I have set it up wrong

below sound setting:
pi@themaker:~/webrtc-test/client $ arecord -L
null
    Discard all samples (playback) or generate zero samples (capture)
dout
dmic_hw
dmic_sv
default
**sysdefault:CARD=sndrpii2scard
    snd_rpi_i2s_card, simple-card_codec_link snd-soc-dummy-dai-0
    Default Audio Device**
dmix:CARD=sndrpii2scard,DEV=0
    snd_rpi_i2s_card, simple-card_codec_link snd-soc-dummy-dai-0
    Direct sample mixing device
dsnoop:CARD=sndrpii2scard,DEV=0
    snd_rpi_i2s_card, simple-card_codec_link snd-soc-dummy-dai-0
    Direct sample snooping device
hw:CARD=sndrpii2scard,DEV=0
    snd_rpi_i2s_card, simple-card_codec_link snd-soc-dummy-dai-0
    Direct hardware device without any conversions
plughw:CARD=sndrpii2scard,DEV=0
    snd_rpi_i2s_card, simple-card_codec_link snd-soc-dummy-dai-0
    Hardware device with all software conversions

**Comment by leeyunjai82** [comment]
1)
`    audio_player = MediaPlayer("sysdefault:CARD=sndrpii2scard", format="alsa", options={'sample_rate':'48000'})
    player = MediaPlayer("/dev/video0", format='v4l2', options={'framerate':'5', 'video_size':'320x240'})

    # configure media
    media = {'audio': False, 'video': False}
    if player and player.video:
        pc.addTrack(player.video)
        media['video'] = True

    #if audio_player and audio_player.audio:
    #    pc.addTrack(audio_player.audio)
    #    print(audio_player.audio)
    #    media['audio'] = True`
>>>    
DATA: {'janus': 'media', 'session_id': 819557512912717, 'sender': 2780613430956354, 'type': 'video', 'receiving': True}
Video OK

2)
`    audio_player = MediaPlayer("sysdefault:CARD=sndrpii2scard", format="alsa", options={'sample_rate':'48000'})
    player = MediaPlayer("/dev/video0", format='v4l2', options={'framerate':'5', 'video_size':'320x240'})

    # configure media
    media = {'audio': False, 'video': False}
    if player and player.video:
        pc.addTrack(player.video)
        media['video'] = True

    if audio_player and audio_player.audio:
        pc.addTrack(audio_player.audio)
        print(audio_player.audio)
        media['audio'] = True`

>>>
DATA: {'janus': 'webrtcup', 'session_id': 7683125598503758, 'sender': 8611677132979134}
DATA: {'janus': 'media', 'session_id': 7683125598503758, 'sender': 8611677132979134, 'type': 'video', 'receiving': True}
DATA: {'janus': 'media', 'session_id': 7683125598503758, 'sender': 8611677132979134, 'type': 'audio', 'receiving': True}
[video4linux2,v4l2 @ 0x8b9b90] Some buffers are still owned by the caller on close.
DATA: {'janus': 'media', 'session_id': 7683125598503758, 'sender': 8611677132979134, 'type': 'audio', 'receiving': False}
DATA: {'janus': 'media', 'session_id': 7683125598503758, 'sender': 8611677132979134, 'type': 'video', 'receiving': False}

Video/Audio no traffic

**Comment by RetroDISTORT** [comment]
I would suggest making a custom MediaStreamTrack subclass.

As an example, this is the one I made. (Mono sound)

I used threads so the program can constantly consume data. This is to prevent overflows
```
from av import AudioFrame
import pyaudio
import av
import numpy as np

from threading import Thread, Event, Lock
from aiortc.mediastreams import MediaStreamTrack


class SystemMic(MediaStreamTrack):
    kind = "audio"
    
    def __init__(self):
        super().__init__()
        
        self.kind         = "audio"
        self.RATE         = 44100
        self.AUDIO_PTIME  = 0.020                                    # 20ms audio packetization
        self.SAMPLES      = int(self.AUDIO_PTIME * self.RATE)
        self.FORMAT       = pyaudio.paInt32
        self.CHANNELS     = 2
        self.CHUNK        = int(self.RATE*self.AUDIO_PTIME)
        self.INDEX        = 0
        self.FORMATAF     = 's16'   #'s32'                           # s32_le
        self.LAYOUT       = 'stereo'
        self.sampleCount  = 0

        self.audio        = pyaudio.PyAudio()
        self.stream       = self.audio.open(format=self.FORMAT, channels=self.CHANNELS, rate=self.RATE,
                                            input=True, input_device_index=self.INDEX,
                                            frames_per_buffer=self.CHUNK)
        #thread
        self.micData          = None
        self.micDataLock      = Lock()
        self.newMicDataEvent  = Event()
        self.newMicDataEvent.clear()
        self.captureThread    = Thread(target=self.capture)
        self.captureThread.start()
        

    def capture(self):
        while True:
            data  = np.fromstring(self.stream.read(self.CHUNK),dtype=np.int32)
            
            with self.micDataLock:
                self.micData = data
                self.newMicDataEvent.set()
    
        
    async def recv(self):
        newMicData = None
            
        self.newMicDataEvent.wait()

        with self.micDataLock:
            data  = self.micData
            data  = (data/2).astype('int32')
            data  = np.array([(data>>16).astype('int16')])
            self.newMicDataEvent.clear()
        
        frame   = av.AudioFrame.from_ndarray(data, self.FORMATAF, layout=self.LAYOUT)
        frame.pts         = self.sampleCount
        frame.rate        = self.RATE
        self.sampleCount += frame.samples

        return frame

    def stop(self):
        super.stop()
        self.captureThread.kill()
```

Before entering your specific microphone data, I would suggest you'd test it first. This code generates an audio file. If your settings are correct, you should hear your mic clearly.
```

import pyaudio
import wave
import numpy as np

FORMAT = pyaudio.paInt32
CHANNELS = 2
RATE = 44100
CHUNK = 4096
RECORD_SECONDS = 5
WAVE_OUTPUT_FILENAME = "file.wav"
device_index = 0
audio = pyaudio.PyAudio()

FORMAT_OUT = pyaudio.paInt16

print("----------------------record device list---------------------")
info = audio.get_host_api_info_by_index(0)
numdevices = info.get('deviceCount')
for i in range(0, numdevices):
    if (audio.get_device_info_by_host_api_device_index(0, i).get('maxInputChannels')) > 0:
        print("Input Device id ", i, " - ", audio.get_device_info_by_host_api_device_index(0, i).get('name'))

print("-------------------------------------------------------------")

index = int(input())
print("recording via index "+str(index))

stream = audio.open(format=FORMAT, channels=CHANNELS,
                    rate=RATE, input=True,input_device_index = index,
                    frames_per_buffer=CHUNK)
print ("recording started")
Recordframes = []

for i in range(0, int(RATE / CHUNK * RECORD_SECONDS)):
    #data = stream.read(CHUNK)
    
    data = np.fromstring(stream.read(CHUNK),dtype=np.int32)
    print(data)
    
    data = (data/2).astype('int32')
    data = (data>>16).astype('int16')
    print(data)
    
    Recordframes.append(data)
print ("recording stopped")

stream.stop_stream()
stream.close()
audio.terminate()

waveFile = wave.open(WAVE_OUTPUT_FILENAME, 'wb')
waveFile.setnchannels(CHANNELS)
waveFile.setsampwidth(audio.get_sample_size(FORMAT_OUT))
waveFile.setframerate(RATE)
waveFile.writeframes(b''.join(Recordframes))
waveFile.close()

```

**Comment by learncodingforweb** [comment]
Thanks @RetroDISTORT , able to send mic audio on webrtc.
